### PR TITLE
KAA-774: [C++ SDK] Add callback for log delivery.

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp [hunk 1]
```diff
@@ -108,51 +108,76 @@ BOOST_AUTO_TEST_CASE(GetAllRecordsTest)
 
     KaaUserLogRecord record;
     record.logdata = testLogData;
-    LogRecordPtr serializedRecord(new LogRecord(record));
+    LogRecord serializedRecord(record);
 
     std::size_t recordCount = 5;
     for (std::size_t i = 0; i < recordCount; ++i) {
-        logStorage.addLogRecord(serializedRecord);
+        logStorage.addLogRecord(std::move(serializedRecord));
```
Comment (forGGe) [inline client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp]: After std::move serializedRecord() is in unspecified state. Are you sure that after moving on first iteration it should be used in following?

File: client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp [hunk 2]
```diff
@@ -108,51 +108,76 @@ BOOST_AUTO_TEST_CASE(GetAllRecordsTest)
...
 
     std::size_t totalSize = 0;
-    for (auto& record : logs.second) {
-        totalSize += record->getSize();
+    for (auto& record : logs.getRecords()) {
+        totalSize += record.getSize();
     }
 
-    BOOST_CHECK_EQUAL(totalSize, recordCount * serializedRecord->getSize());
+    BOOST_CHECK_EQUAL(totalSize, recordCount * serializedRecord.getSize());
```
Comment (forGGe) [inline client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp]: Object was moved previously, but here it is used again without preceding assignment.

File: client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp (hunk 1 above)
Comment (forGGe) [inline client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp]: Sorry for that, objects defined in STL are in unspecified state after moving. 
But a question remains about correctly used move-semantics in this case.

File: client/client-multi/client-cpp/test/impl/log/MemoryLogStorageTest.cpp [hunk 3]
```diff
@@ -77,156 +73,184 @@ BOOST_AUTO_TEST_CASE(AddRecordsAndCheckStatusTest)
...
     std::size_t logRecordCount = 10;
-    auto serializedLogRecord = createSerializedLogRecord();
-    std::size_t maxLogStorageSize = logRecordCount * serializedLogRecord->getSize();
+    std::size_t serializedLogSize = createSerializedLogRecord().getSize();
+    std::size_t maxLogStorageSize = logRecordCount * serializedLogSize;
     float percentToDelete = 51.1;
 
-    MemoryLogStorage logStorage(maxLogStorageSize, percentToDelete);
+    MemoryLogStorage logStorage(maxLogStorageSize, percentToDelete, serializedLogSize * logRecordCount / 2);
     for (std::size_t i = 1; i <= logRecordCount; ++i) {
```
Comment (forGGe) [inline client/client-multi/client-cpp/test/impl/log/MemoryLogStorageTest.cpp:255]: Strange counter.

File: client/client-multi/client-cpp/test/impl/log/LogCollectorTest.cpp [hunk 4]
```diff
@@ -324,28 +311,29 @@ BOOST_AUTO_TEST_CASE(TimeoutDetectionTest)
...
 
-    ILogStorage::RecordBlock block{ createSerializedLogRecord(),
+    std::list<LogRecord> logs{ createSerializedLogRecord(),
                                     createSerializedLogRecord(),
                                     createSerializedLogRecord(),
                                     createSerializedLogRecord()
                                   };
 
-    ILogStorage::RecordPack recordPack(1, block);
+    LogBucket bucket(1, (logs));
```
Comment (forGGe) [inline client/client-multi/client-cpp/test/impl/log/LogCollectorTest.cpp]: Guess that std::move is forgotten here.

**Comment by forGGe** [comment]
Reviewed.

File: client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp (hunk 1 above)
Comment (ghost) [inline client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp]: Fixed.

File: client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp (hunk 2 above)
Comment (ghost) [inline client/client-multi/client-cpp/test/impl/log/SQLiteDBLogStorageTest.cpp]: Fixed.

File: client/client-multi/client-cpp/test/impl/log/MemoryLogStorageTest.cpp (hunk 3 above)
Comment (ghost) [inline client/client-multi/client-cpp/test/impl/log/MemoryLogStorageTest.cpp:255]: Fixed.

File: client/client-multi/client-cpp/test/impl/log/LogCollectorTest.cpp (hunk 4 above)
Comment (ghost) [inline client/client-multi/client-cpp/test/impl/log/LogCollectorTest.cpp]: Fixed.

File: client/client-multi/client-cpp/impl/log/LogCollector.cpp [hunk 5]
```diff
@@ -264,55 +264,64 @@ std::shared_ptr<LogSyncRequest> LogCollector::getLogUploadRequest()
...
+    request->requestId = bucket.getBucketId();
 
-        request->logEntries.set_array(std::move(logs));
-        addDeliveryTimeout(request->requestId);
+    std::vector<LogEntry> logsToSend;
+    logsToSend.resize(bucket.getRecords().size());
+
+    std::size_t i = 0;
+    for (auto& record : bucket.getRecords()) {
+        logsToSend[i++].data = std::move(record.getRvalueData());
```
Comment (ohamov) [inline client/client-multi/client-cpp/impl/log/LogCollector.cpp:283]: std::move is redundant here. Please remove it.

**Comment by forGGe** [comment]
Looks good!

**Pull request merged**